### PR TITLE
Implement Kumaraswamy Distribution

### PR DESCRIPTION
**Pull request by vishwakftw**
This closes the proposal in https://github.com/probtorch/pytorch/issues/140.

- [x] Distribution
- [ ] Tests
~~KL Divergence~~

cc: @alicanb @fritzo

### CODE REVIEW

File: torch/distributions/kumaraswamy.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    has_rsample = True
+
+    def __init__(self, concentration1, concentration0, validate_args=None):
+        self.concentration1, self.concentration0 = broadcast_all(concentration1, concentration0)
+        base_dist = Uniform(torch.zeros_likes(self.concentration0), torch.ones_like(self.concentration0))
+        transforms = [AffineTransform(loc=1, scale=-torch.ones_like(self.concentration1)),
+                      PowerTransform(exponent=self.concentration1.reciprocal()),
+                      AffineTransform(loc=1, scale=-torch.ones_like(self.concentration0)),
+                      PowerTransform(exponent=self.concentration0.reciprocal())]
+        super(Pareto, self).__init__(base_dist, transforms, validate_args=validate_args)
```
Comment (fritzo) [inline torch/distributions/kumaraswamy.py]: Replace `Pareto` with `Kumaraswamy`

**Review by fritzo — COMMENTED** [review]
Tests?

Also, how do you plan to test `.mean` and `.variance`? Should we add monte carlo test similar to the `.entropy` test?

File: torch/distributions/kumaraswamy.py (hunk 1 above)
Comment (vishwakftw) [reply torch/distributions/kumaraswamy.py]: Oops, I am sorry.

**Comment by vishwakftw** [comment]
@fritzo I am working on the tests now. Unfortunately, SciPy doesn't have the Kumaraswamy distribution in its distribution suite, which means the tests have to be Monte-Carlo based.

File: torch/distributions/kumaraswamy.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+         0.1729
+        [torch.FloatTensor of size (1,)]
+
+    Args:
+        concentration1 (float or Tensor): 1st concentration parameter of the distribution
+            (often referred to as alpha)
+        concentration0 (float or Tensor): 2nd concentration parameter of the distribution
+            (often referred to as beta)
+    """
+    params = {'concentration1': constraints.positive, 'concentration0': constraints.positive}
```
Comment (fritzo) [inline torch/distributions/kumaraswamy.py]: FYI `.params` was just renamed to `.arg_constraints` to avoid confusion with `.parameter` as used elsewhere in PyTorch. See https://github.com/pytorch/pytorch/pull/5989

**Comment by vishwakftw** [comment]
Closing due to inactivity.

**Comment by vishwakftw** [comment]
@alicanb I have cleared up the PR (basically removed merge conflicts) after reopening. Mind taking a look?

**Review by alicanb — CHANGES_REQUESTED** [review]
One additional test we can add is Kumaraswamy(a,1)=Beta(a,1) and Kumaraswamy(1,b)=Beta(1,b). I will look at the entropy later today and get back to you.

File: torch/distributions/kumaraswamy.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    has_rsample = True
+
+    def __init__(self, concentration1, concentration0, validate_args=None):
+        self.concentration1, self.concentration0 = broadcast_all(concentration1, concentration0)
+        finfo = _finfo(self.concentration0)
+        base_dist = Uniform(torch.full_like(self.concentration0, finfo.tiny),
+                            torch.full_like(self.concentration0, 1 - finfo.eps))
+        transforms = [AffineTransform(loc=1., scale=-torch.ones_like(self.concentration1)),
+                      PowerTransform(exponent=self.concentration0.reciprocal()),
+                      AffineTransform(loc=1., scale=-torch.ones_like(self.concentration0)),
```
Comment (vishwakftw) [inline torch/distributions/kumaraswamy.py]: It's so bizarre why this is happening: the base distribution samples points in the open interval (0, 1), and the first two transforms work correctly by maintaining this range. However, this `AffineTransform` seems to make certain values close to 0 = 1, which is causing the entropy test failure.

Any thoughts on how to mitigate this issue?

Comment (fritzo) [reply torch/distributions/kumaraswamy.py]: Change the `.tiny` to `.eps`. This is because `eps` is around 1e-8 but tiny is around 1e-38. Floating point representation can measure very tiny differences near 0, but only 1e-8 differences near 1.

Comment (vishwakftw) [reply torch/distributions/kumaraswamy.py]: I tried this earlier, and it didn't seem to work. Even then, it was the same transform causing the problem too.

Comment (fritzo) [reply torch/distributions/kumaraswamy.py]: Hmm you could try changing it to `self.concentration * eps`, to ensure it's still measurable after the `PowerTransform`?

Comment (vishwakftw) [reply torch/distributions/kumaraswamy.py]:
```diff
-        base_dist = Uniform(torch.full_like(self.concentration0, finfo.tiny),
+        base_dist = Uniform(torch.full_like(self.concentration0, finfo.eps),
                             torch.full_like(self.concentration0, 1 - finfo.eps))
         transforms = [AffineTransform(loc=1., scale=-torch.ones_like(self.concentration1)),
                       PowerTransform(exponent=self.concentration0.reciprocal()),
-                      AffineTransform(loc=1., scale=-torch.ones_like(self.concentration0)),
+                      AffineTransform(loc=1., scale=-torch.full_like(self.concentration0, 1 - finfo.eps)),
```

Sorry @fritzo , even this doesn't seem to help? Is the diff what you had in mind?

Comment (vishwakftw) [reply torch/distributions/kumaraswamy.py]: `eps` is about 1e-08, and if we were to assume `self.concentration0` = 1e08 (hypothetically, of course, but still a valid parameter value), then the base distribution will get affected, won't it?

Comment (vishwakftw) [reply torch/distributions/kumaraswamy.py]: However, even that doesn't seem to help.

Comment (fritzo) [reply torch/distributions/kumaraswamy.py]: 1. I think we can safely ignore extreme concentration like `concentration0=1-e8`, or to require users to use double precision in extreme cases.
2. For non-extreme cases, it shouldn't matter that the base distribution is affected, since `PowTransform` will map the smallest value back down to `eps`. Of course there will be mass missing at 0, but we've already decided to ignore mass at discrete points.

Comment (vishwakftw) [reply torch/distributions/kumaraswamy.py]: Thanks @fritzo , but your solution doesn't seem to work. This is how I am finding out the bug:
```diff
                 expected = -dist.log_prob(x).mean(0)
                 ignore = (expected == inf)
                 expected[ignore] = actual[ignore]
+                if torch.any(torch.isnan(expected)):
+                    x = dist.base_dist.sample(sample_shape=(60000,))
+                    for trans in dist.transforms:
+                        x = trans(x)
+                        print(trans)
+                        print(torch.any(x >= 1))
+                        print(torch.any(x <= 0))
```

This is the captured output by `pytest`:
```
AffineTransform()
tensor(0, dtype=torch.uint8)
tensor(0, dtype=torch.uint8)
PowerTransform()
tensor(0, dtype=torch.uint8)
tensor(0, dtype=torch.uint8)
AffineTransform()
tensor(1, dtype=torch.uint8)
tensor(0, dtype=torch.uint8)
PowerTransform()
tensor(1, dtype=torch.uint8)
tensor(0, dtype=torch.uint8)
```

**Review by fritzo — COMMENTED** [review]
BTW What applications do you have in mind for the `Kumaraswami` distribution? As I understand it was mainly the choice for reparametrized `Beta`-like distribution prior to availability of reparametrized `Beta`. Is it still useful?

File: torch/distributions/kumaraswamy.py (hunk 3 above)
Comment (fritzo) [reply torch/distributions/kumaraswamy.py]: Ok, why don't you compute the first-order scaling of `eps` that is needed to make it through all transforms. It should involve `concentration0` and `concentration1`.

**Comment by vishwakftw** [comment]
Perhaps @alicanb would be the best person to answer this question. I re-opened this on account of a feature request in `pytorch/pytorch` @ https://github.com/pytorch/pytorch/issues/11937

**Comment by bastings** [comment]
Hi, @wilkeraziz and I proposed adding this distribution in the issue. Beta and Kumaraswamy both give support to variables in the open interval (0, 1), but they are not the same distribution (for example, a Kumaraswamy is not an exponential family), thus having one does not justify leaving the other out. Kumaraswamy variables are simpler to reparameterise because they can be seen as a bijective transformation of a uniform variable (which means we can reparameterise gradients). This is not the case for Beta variables, in fact, the most effective techniques are rather complex and typically lead to higher variance (check https://arxiv.org/abs/1610.05683 and https://arxiv.org/abs/1603.00788 and https://arxiv.org/abs/1610.02287 for alternatives that are compatible with Beta and Dirichlet, for example).

Short answer: having the Kumaraswamy would be great and we would use it in our research, probably so would others. 

Thanks for the work so far @vishwakftw @fritzo ! Is the entropy instability the only remaining issue?

Q: The Kuma I wanted to PR was a direct bijection of the Uniform base distribution (similar to how it's done in TF Probability). Could that solve the instability issues?

**Comment by vishwakftw** [comment]
@bastings Thank you for the clarification. 

The implementation of the `Kumaraswamy` distribution in this PR is also based on bijections (called `Transforms` here). There seems to be some issues with the transform itself, which I haven't been able to figure out yet.

**Comment by fritzo** [comment]
@bastings Don't forget https://arxiv.org/abs/1806.01851 which describes the reparametrization used by PyTorch's Gamma, Beta, and Dirichlet.

**Comment by bastings** [comment]
> @bastings Don't forget https://arxiv.org/abs/1806.01851 which describes the reparametrization used by PyTorch's Gamma, Beta, and Dirichlet.

Thanks for the pointer :-)

**Comment by bastings** [comment]
@vishwakftw @alicanb @fritzo is this still failing in current master? maybe we can have a fresh look at it?

**Comment by keithmoffat** [comment]
Just wanted to check if the Kumaraswamy distribution has been implemented yet? If not, Id like to give it a +1 :)

**Comment by RylanSchaeffer** [comment]
I'm also looking for an update

**Comment by sdaulton** [comment]
+1. Is anyone working on this?

File: torch/distributions/kumaraswamy.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+    @property
+    def variance(self):
+        return _moments(self.concentration1, self.concentration0, 2) - torch.pow(self.mean, 2)
+
+    def entropy(self):
+        t1 = (1 - self.concentration1.reciprocal())
+        t0 = (1 - self.concentration0.reciprocal())
+        H0 = torch.digamma(self.concentration0 + 1) + euler_constant
+        return t1 + t0 * H0 + torch.log(self.concentration1) + torch.log(self.concentration0)
```
Comment (sdaulton) [inline torch/distributions/kumaraswamy.py:67]: This should be `t0 + t1 * H0 + ...`. From https://en.wikipedia.org/wiki/Kumaraswamy_distribution, we want to compute H_b. 

Aside: concentration1 -> alpha, concentration0 -> beta is rather confusion. Can we rename these for be clearer?

**Review by sdaulton — CHANGES_REQUESTED** [review]

File: torch/distributions/kumaraswamy.py (hunk 4 above)
Comment (fritzo) [reply torch/distributions/kumaraswamy.py:67]: These are the names from tensorflow probability and have been used throughout `torch.distributions`. We have aimed for compatibility and consistency over subjective clarity.

**Comment by sdaulton** [comment]
@vishwakftw are you planning to iterate on this PR and land it or is this abandoned? Should I commandeer it?

**Comment by vishwakftw** [comment]
I could wrap it up, but this is in `probtorch`. @fritzo should I create one upstream?

**Comment by fritzo** [comment]
@vishwakftw sure, create a PR upstream in pytorch.

**Comment by vishwakftw** [comment]
Sure, I'll do that first thing this afternoon. Thanks @fritzo @sdaulton.

**Comment by vishwakftw** [comment]
Apologies for the delay, I'll create a PR later this evening (too many constraints at the moment).

**Comment by vishwakftw** [comment]
I'm facing some tests failures, which is why I haven't created a PR upstream. All discussions about this can be seen above.

**Comment by vishwakftw** [comment]
Merged upstream, closing stale PR.
